internal/controllers/controllers.go

```diff
@@ -0,0 +1,83 @@
+package controllers
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strconv"
+
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const (
+	// This label is set on all dynamic objects, to limit caches.
+	DynamicCacheLabel = "package-operator.run/cache"
+	// Revision annotations holds a revision generation number to order ObjectSets.
+	RevisionAnnotation = "package-operator.run/revision"
+)
+
+func EnsureCommonFinalizer(
+	ctx context.Context, obj client.Object,
+	c client.Client, cacheFinalizer string,
+) error {
+	if !controllerutil.ContainsFinalizer(obj, cacheFinalizer) {
+		controllerutil.AddFinalizer(obj, cacheFinalizer)
+		if err := c.Update(ctx, obj); err != nil {
+			return fmt.Errorf("adding finalizer: %w", err)
+		}
+	}
+	return nil
+}
+
+type dynamicWatchFreer interface {
+	Free(ctx context.Context, obj client.Object) error
+}
+
+func HandleCommonDeletion(
+	ctx context.Context, obj client.Object,
+	c client.Client, dw dynamicWatchFreer,
+	cacheFinalizer string,
```

Comment:
Does `cacheFinalizer` really need to be a parameter? Isn't it static?

Reply:
It can be static, but I found it best practice to name it by controller waiting on the finalizer:
e.g. `objectset.package-operator.run/cache`, `objectdeployment.package-operator.run/cache` etc.
If we ever refactor code so multiple finalizers are present on the same object, it's way easier to tell them apart and ensure that we are not removing the wrong finalizer.

Reply:
Ohh okay that makes sense if there is a situation with two controllers adding the cache finalizer to one object, but I am kind of unclear what the point of that would be. Couldn't other issues also arise if there were two cache finalizers? From what I see, running the cacheFinalizer calls `FreeCacheAndFinalizer` which doesn't depend on the controller, only the object. So after the first cache finalizer is called the cache will already be freed, but it has a finalizer that says it needs to free the cache again.

Reply:
Freeing the cache multiple times is no problem.
But removing the finalizer before the cache is cleaned up is an issue, because if freeing the cache failed, we cannot add the finalizer again.